eth/fetcher/block_fetcher.go

```diff
@@ -461,15 +462,28 @@ func (f *BlockFetcher) loop() {
 
 				// Create a closure of the fetch and schedule in on a new thread
 				fetchHeader, hashes := f.fetching[hashes[0]].fetchHeader, hashes
-				go func() {
+				go func(peer string) {
 					if f.fetchingHook != nil {
 						f.fetchingHook(hashes)
 					}
 					for _, hash := range hashes {
 						headerFetchMeter.Mark(1)
-						fetchHeader(hash) // Suboptimal, but protocol doesn't allow batch header retrievals
+						go func(hash common.Hash) {
```

Comment:
This function now starts peer * hashes go routines instead of peer routines. I think that's by design, just wanted to make sure that this won't become a problem

Reply:
Myeah, that's a good observation, alas delliberate.

Previously `fetchHeader` just sent a remote request and the reply came back from the outside from the handler on a separate goroutine.

With the request id style blocking APIs, I need something to wait for the reply to actually arrive. A better way would be to make a demultiplexer that listens on a sink channel and keeps consuming inbound responses. It requires a bit of a heavier internal rewrite that I didn't want to focus on yet. Though the question that's valid is whether that part should be done before a merge.

Reply:
In practical terms, a new goroutine per request should be ok-ish, as we're only launching on request per missing data, so the only way to blow up the goroutine count is to blow up the block announcements; but that's limited specifically to avoid blowing up memory usage with spam. So we're still kind of bound by the old DoS protections, just it's +1 goroutine load.